EIPS/eip-4844.md

```diff
@@ -0,0 +1,558 @@
+
+### Helpers
+
+Converts a blob to its corresponding KZG point:
+
+```python
+def blob_to_kzg(blob: Vector[BLSFieldElement, FIELD_ELEMENTS_PER_BLOB]) -> KZGCommitment:
+    computed_kzg = bls.Z1
+    for value, point_kzg in zip(tx.blob, KZG_SETUP_LAGRANGE):
+        assert value < BLS_MODULUS
+        computed_kzg = bls.add(
+            computed_kzg,
+            bls.multiply(point_kzg, value)
+        )
+    return computed_kzg
+```
+
+Converts a KZG point into a versioned hash:
+
+```python
+def kzg_to_versioned_hash(kzg: KZGCommitment) -> VersionedHash:
+    return BLOB_COMMITMENT_VERSION_KZG + hash(kzg)[1:]
+```
+
+Verifies a KZG evaluation proof:
+
+```python
+def verify_kzg_proof(polynomial_kzg: KZGCommitment,
+                     x: BLSFieldElement,
+                     y: BLSFieldElement,
+                     quotient_kzg: KZGCommitment):
+    # Verify: P - y = Q * (X - x)
+    X_minus_x = bls.add(KZG_SETUP_G2[1], bls.multiply(bls.G2, BLS_MODULUS - x))
+    P_minus_y = bls.add(polynomial_kzg, bls.multiply(bls.G1, BLS_MODULUS - y))
+    return bls.pairing_check([
+        [P_minus_y, bls.neg(bls.G2)],
+        [quotient_kzg, X_minus_x]
+    ])
+```
+
+Approximates `2 ** (numerator / denominator)`, with the simplest possible approximation that is continuous and has a continuous derivative:
+
+```python
+def fake_exponential(numerator: int, denominator: int) -> int:
+    cofactor = 2 ** (numerator // denominator)
+    fractional = numerator % denominator
+    return cofactor + (
+        fractional * cofactor * 2 +
+        (fractional ** 2 * cofactor) // denominator
+    ) // (denominator * 3)
+```
+
+### New transaction type
+
+We introduce a new [EIP-2718](./eip-2718.md) transaction type,
+with the format being the single byte `BLOB_TX_TYPE` followed by an SSZ encoding of the 
+`SignedBlobTransaction` container comprising the transaction contents:
+
+```python
+class SignedBlobTransaction(Container):
+    message: BlobTransaction
+    signature: ECDSASignature
+
+class BlobTransaction(Container):
+    chain_id: uint256
+    nonce: uint64
+    priority_fee_per_gas: uint256
+    max_basefee_per_gas: uint256
+    gas: uint64
+    to: Union[None, Address] # Address = Bytes20
+    value: uint256
+    data: ByteList[MAX_CALLDATA_SIZE]
+    access_list: List[AccessTuple, MAX_ACCESS_LIST_SIZE]
+    blob_versioned_hashes: List[VersionedHash, MAX_VERSIONED_HASHES_LIST_SIZE]
+
+class AccessTuple(Container):
+    address: Address # Bytes20
+    storage_keys: List[Hash, MAX_ACCESS_LIST_STORAGE_KEYS]
+
+class ECDSASignature(Container):
+    v: uint8
+    r: uint256
+    s: uint256
+```
+
+The `priority_fee_per_gas` and `max_basefee_per_gas` fields follow [EIP-1559](./eip-1559.md) semantics,
+and `access_list` as in [`EIP-2930`](./eip-2930.md).
+
+[`EIP-2718`](./eip-2718.md) is extended with a "wrapper data", the typed transaction can be encoded in two forms, dependent on the context:
```

Comment:
If this is to be a general purpose extension to 2718, we might want to strip it out into a separate EIP

Reply:
Before creating a whole new EIP I want to get feedback in ACD first (there may be different ways we split this EIP).